tests/integ/modin/frame/test_cache_result.py

```diff
@@ -0,0 +1,269 @@
+        )
+    with SqlCounter(query_count=1):
+        assert_empty_snowpark_pandas_equals_to_pandas(cached_snow_df, native_df)
+
+    native_df = native_pd.DataFrame(columns=["A", "B", "C"], index=[0, 1, 2])
+    snow_df = pd.DataFrame(native_df)
+    snow_df_copy = snow_df.copy(deep=True)
+    with SqlCounter(query_count=1):
+        cached_snow_df = cache_and_return_df(snow_df, inplace)
+    with SqlCounter(query_count=2):
+        assert_frame_equal(
+            snow_df_copy.to_pandas(), cached_snow_df.to_pandas(), check_index_type=False
+        )
+    with SqlCounter(query_count=1):
+        assert_snowpark_pandas_equals_to_pandas_without_dtypecheck(
+            cached_snow_df, native_df
+        )
+
+
+@pytest.mark.parametrize("inplace", [True, False])
+def test_cache_result_dataframe_complex_correctness(
+    time_index_string_column_data,
+    inplace,
+):
+    df_data, kwargs = time_index_string_column_data
+    snow_df, native_df = create_test_dfs(df_data, **kwargs)
+
+    snow_df = snow_df.resample("2H").mean()
+    snow_df_copy = snow_df.copy(deep=True)
+    with SqlCounter(query_count=1):
+        cached_snow_df = cache_and_return_df(snow_df, inplace)
+    with SqlCounter(query_count=2):
+        assert_frame_equal(
+            snow_df_copy.to_pandas(), cached_snow_df.to_pandas(), check_index_type=False
+        )
+    native_df = native_df.resample("2H").mean()
+
+    cached_snow_df = cached_snow_df.set_index("b", drop=False)
+    native_df = native_df.set_index("b", drop=False)
+    with SqlCounter(query_count=1):
+        assert_snowpark_pandas_equals_to_pandas_without_dtypecheck(
+            cached_snow_df, native_df, check_freq=False
+        )
+
+
+@pytest.mark.parametrize("inplace", [True, False])
+class TestCacheResultReducesQueryCount:
+    def test_cache_result_simple(self, inplace):
+        snow_df = pd.concat([pd.DataFrame([range(i, i + 5)]) for i in range(0, 150, 5)])
+        native_df = perform_chained_operations(
+            native_pd.DataFrame(np.arange(150).reshape((30, 5))), native_pd
+        )
+        with SqlCounter(query_count=1, union_count=299):
+            snow_df = perform_chained_operations(snow_df, pd)
+            assert_snowpark_pandas_equals_to_pandas_without_dtypecheck(
+                snow_df, native_df
+            )
+
+        with SqlCounter(query_count=1, union_count=29):
+            snow_df = pd.concat(
+                [pd.DataFrame([range(i, i + 5)]) for i in range(0, 150, 5)]
+            )
+            cached_snow_df = cache_and_return_df(snow_df, inplace)
+
+        with SqlCounter(query_count=1, union_count=9):
+            cached_snow_df = perform_chained_operations(cached_snow_df, pd)
+            cached_snow_df.to_pandas()
+
+        with SqlCounter(query_count=1):
+            assert_snowpark_pandas_equals_to_pandas_without_dtypecheck(
+                cached_snow_df, native_df
+            )
```

Comment:
```suggestion
            assert_snowpark_pandas_equals_to_pandas_without_dtypecheck(
                cached_snow_df, native_df
            )
```

nit: this is clearer than the `cached_snow_df.to_pandas()` whose result isn't used anywhere.